Sources/NIOPosix/NonBlockingFileIO.swift

```diff
@@ -712,3 +730,346 @@ public struct NIODirectoryEntry: Hashable {
+                return Int64(eof)
+            }
+        }
+    }
+
+    /// Write `buffer` to `fileHandle` in ``NonBlockingFileIO``'s private thread pool which is separate from any ``EventLoop`` thread.
+    ///
+    /// - parameters:
+    ///   - fileHandle: The ``NIOFileHandle`` to write to.
+    ///   - buffer: The ``ByteBuffer`` to write.
+    public func write(fileHandle: NIOFileHandle,
+                      buffer: ByteBuffer) async throws {
+        return try await self.write0(fileHandle: fileHandle, toOffset: nil, buffer: buffer)
+    }
+
+    /// Write `buffer` starting from `toOffset` to `fileHandle` in ``NonBlockingFileIO``'s private thread pool which is separate from any ``EventLoop`` thread.
+    ///
+    /// - parameters:
+    ///   - fileHandle: The ``NIOFileHandle`` to write to.
+    ///   - toOffset: The file offset to write to.
+    ///   - buffer: The ``ByteBuffer`` to write.
+    public func write(fileHandle: NIOFileHandle,
+                      toOffset: Int64,
+                      buffer: ByteBuffer) async throws {
+        return try await self.write0(fileHandle: fileHandle, toOffset: toOffset, buffer: buffer)
+    }
+
+    private func write0(fileHandle: NIOFileHandle,
+                        toOffset: Int64?,
+                        buffer: ByteBuffer) async throws {
+        let byteCount = buffer.readableBytes
+
+        guard byteCount > 0 else {
+            return
+        }
+
+        return try await self.threadPool.runIfActive {
+            try self.writeSync(fileHandle: fileHandle, byteCount: byteCount, toOffset: toOffset, buffer: buffer)
+        }
+    }
+
+    /// Open the file at `path` for reading on a private thread pool which is separate from any ``EventLoop`` thread.
+    ///
+    /// This function will return (a future) of the ``NIOFileHandle`` associated with the file opened and a ``FileRegion``
+    /// comprising of the whole file. The caller must close the returned ``NIOFileHandle`` when it's no longer needed.
+    ///
+    /// - note: The reason this returns the ``NIOFileHandle`` and the ``FileRegion`` is that both the opening of a file as well as the querying of its size are blocking.
+    ///
+    /// - parameters:
+    ///     - path: The path of the file to be opened for reading.
+    /// - returns: the ``NIOFileHandle`` and the ``FileRegion`` comprising the whole file.
+    public func withOpenFile<Result>(path: String, _ body: (NIOFileHandle, FileRegion) async throws -> Result) async throws -> Result {
+        let file = try await self.threadPool.runIfActive {
+            let fh = try NIOFileHandle(path: path)
+            do {
+                let fr = try FileRegion(fileHandle: fh)
+                return UnsafeTransfer((handle: fh, region: fr))
+            } catch {
+                _ = try? fh.close()
+                throw error
+            }
+        }
+        defer {
+            try? file.wrappedValue.handle.close()
```

Comment:
Currently this is a `try?`, would you prefer to throw an error from the close?

Reply:
We should throw the error; there's not a lot the user can do about `EBADF` and presumably we handle `EINTR` but it may be `EIO` which the caller may want to know about.